### PR TITLE
Feature/background size

### PR DESCRIPTION
**Pull request by simonberger**

### CODE REVIEW

File: src/Renderer/AbstractRenderer.php [hunk 1]
```diff
@@ -124,6 +166,28 @@ protected function _background_image($url, $x, $y, $width, $height, $style)
         $bg_width = round((float)($width * $dpi) / 72);
         $bg_height = round((float)($height * $dpi) / 72);
 
+        $_background_size = $style->get_background_size();
+        $is_calculated_background = ! is_array($_background_size);
```
[the remark this answers is not in the document]
Comment (bsweeney) [reply src/Renderer/AbstractRenderer.php]: Maybe one of my issues is the name of the variable. I don't think it's sufficiently conveys what we want to know. Maybe `$is_keyword` or something like that?

File: src/Renderer/AbstractRenderer.php [hunk 2]
```diff
@@ -124,6 +166,28 @@ protected function _background_image($url, $x, $y, $width, $height, $style)
...
+        list($img_w, $img_h) = $this->_resize_background_image(
+            $img_w,
+            $img_h,
+            $bg_width,
+            $bg_height,
+            $_background_size
+        );
+
+        if (Helpers::is_percent($img_w)) {
```
Comment (bsweeney) [inline src/Renderer/AbstractRenderer.php]: One of the calculations we need to accommodate, in addition to a percentage and a unit value, is when one of the sides is set to "auto". In that case the side specified as auto should be calculated based on the ratio compared to the length of the other side. If both sides are auto we just use the image dimensions.

**Comment by bsweeney** [comment]
ref https://drafts.csswg.org/css-backgrounds-3/#background-size

**Comment by simonberger** [comment]
> This is a good start. A few tweaks and I think it'll be very close to ready.
> 
> I would like to think on the overall logic a bit more. I feel like some of these calculations could be moved inside the style class setter for better performance around computed values (if possible).

Thanks for your review. Yes the calculation and auto handling is what needs most changes. The current state just adds the calculation in a random way. I already have some improvements idea. You are right they should be movable to the style class when we get pixel/point values there.
I'll do it end of next week and planning it for the release after 0.8.5.

File: src/Renderer/AbstractRenderer.php [hunk 3]
```diff
@@ -124,6 +166,28 @@ protected function _background_image($url, $x, $y, $width, $height, $style)
         $bg_width = round((float)($width * $dpi) / 72);
         $bg_height = round((float)($height * $dpi) / 72);
 
+        $_background_size = $style->get_background_size();
```
[the remark this answers is not in the document]
Comment (simonberger) [reply src/Renderer/AbstractRenderer.php]: changed

File: src/Renderer/AbstractRenderer.php (hunk 2 above)
Comment (simonberger) [reply src/Renderer/AbstractRenderer.php]: Auto handling / calculation implemented now

File: src/Renderer/AbstractRenderer.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (simonberger) [reply src/Renderer/AbstractRenderer.php]: Moved calculations to style class and resize method.

File: src/Css/Style.php [hunk 4]
```diff
@@ -1288,6 +1290,38 @@ function get_background_position()
...
+            case "cover":
+                return 'cover';
+            case "contain":
+                return 'contain';
+            default:
+                break;
+        }
+
+        if (!isset($result[1])) {
+            $result[1] = '';
```
[the remark this answers is not in the document]
Comment (simonberger) [reply src/Css/Style.php]: Changed the way default and auto values are handled. I think it is a lot clearer now beside fixing many cases.

**Comment by bsweeney** [comment]
I haven't done a final review of the code but initial testing of the PDF render and this looks great!

**Comment by simonberger** [comment]
Okay thanks. I already did a lot of tests. I'll do some more next week, think about adding some unit tests and rebase before a merge.

**Comment by UtechtDustin** [comment]
When is the next release with that fix ?

**Comment by bsweeney** [comment]
We don't work off of a set release schedule do to resource constraints. But I expect the next release will come sometime in the next month.

**Comment by holema** [comment]
Hello,
will be there a realease of this feature in the masterbranch?

Best regards

**Comment by bsweeney** [comment]
Yes, trying to get the next release out by the end of the month.

**Pull request merged**